### PR TITLE
Get rid of custom bufio package.

### PR DESCRIPTION
**Pull request by vmihailenco**
```
benchmark                         old ns/op     new ns/op     delta
BenchmarkParseReplyStatus-4       104           106           +1.92%
BenchmarkParseReplyInt-4          139           142           +2.16%
BenchmarkParseReplyError-4        202           208           +2.97%
BenchmarkParseReplyString-4       180           202           +12.22%
BenchmarkParseReplySlice-4        889           949           +6.75%
BenchmarkAppendArgs-4             228           228           +0.00%
BenchmarkRedisClusterPing-4       11370         11477         +0.94%
BenchmarkPool-4                   215           213           -0.93%
BenchmarkRedisPing-4              7073          6933          -1.98%
BenchmarkRedisSet-4               20163         19978         -0.92%
BenchmarkRedisGetNil-4            7055          7192          +1.94%
BenchmarkRedisSetGet64Bytes-4     15489         15449         -0.26%
BenchmarkRedisSetGet1KB-4         19774         18265         -7.63%
BenchmarkRedisSetGet10KB-4        69657         54929         -21.14%
BenchmarkRedisSetGet1MB-4         3493441       2901501       -16.94%
BenchmarkRedisSetGetBytes-4       53624         40511         -24.45%
BenchmarkRedisMGet-4              7331          7449          +1.61%
BenchmarkSetExpire-4              15758         15555         -1.29%
BenchmarkPipeline-4               9800          9720          -0.82%
BenchmarkZAdd-4                   8150          8394          +2.99%

benchmark                         old allocs     new allocs     delta
BenchmarkParseReplyStatus-4       1              1              +0.00%
BenchmarkParseReplyInt-4          2              2              +0.00%
BenchmarkParseReplyError-4        2              2              +0.00%
BenchmarkParseReplyString-4       2              2              +0.00%
BenchmarkParseReplySlice-4        11             11             +0.00%
BenchmarkAppendArgs-4             0              0              +0.00%
BenchmarkRedisClusterPing-4       5              5              +0.00%
BenchmarkPool-4                   0              0              +0.00%
BenchmarkRedisPing-4              5              5              +0.00%
BenchmarkRedisSet-4               8              8              +0.00%
BenchmarkRedisGetNil-4            5              5              +0.00%
BenchmarkRedisSetGet64Bytes-4     16             15             -6.25%
BenchmarkRedisSetGet1KB-4         16             15             -6.25%
BenchmarkRedisSetGet10KB-4        17             16             -5.88%
BenchmarkRedisSetGet1MB-4         37             31             -16.22%
BenchmarkRedisSetGetBytes-4       16             15             -6.25%
BenchmarkRedisMGet-4              16             16             +0.00%
BenchmarkSetExpire-4              15             15             +0.00%
BenchmarkPipeline-4               20             19             -5.00%
BenchmarkZAdd-4                   10             10             +0.00%

benchmark                         old bytes     new bytes     delta
BenchmarkParseReplyStatus-4       32            32            +0.00%
BenchmarkParseReplyInt-4          32            32            +0.00%
BenchmarkParseReplyError-4        32            32            +0.00%
BenchmarkParseReplyString-4       48            48            +0.00%
BenchmarkParseReplySlice-4        240           240           +0.00%
BenchmarkAppendArgs-4             0             0             +0.00%
BenchmarkRedisClusterPing-4       160           161           +0.63%
BenchmarkPool-4                   0             0             +0.00%
BenchmarkRedisPing-4              160           160           +0.00%
BenchmarkRedisSet-4               10754         10754         +0.00%
BenchmarkRedisGetNil-4            176           176           +0.00%
BenchmarkRedisSetGet64Bytes-4     720           592           -17.78%
BenchmarkRedisSetGet1KB-4         3664          2512          -31.44%
BenchmarkRedisSetGet10KB-4        42464         31964         -24.73%
BenchmarkRedisSetGet1MB-4         4212478       3163646       -24.90%
BenchmarkRedisSetGetBytes-4       31979         21477         -32.84%
BenchmarkRedisMGet-4              432           432           +0.00%
BenchmarkSetExpire-4              480           480           +0.00%
BenchmarkPipeline-4               992           864           -12.90%
BenchmarkZAdd-4                   288           288           +0.00%
```

### CODE REVIEW

**Comment by vmihailenco** [comment]
I can't explain slowdown of `BenchmarkParseReplyString`...

**Comment by vmihailenco** [comment]
On the 2nd though it must be the cost of `copy`.

File: conn.go [hunk 1]
```diff
@@ -102,3 +103,16 @@ func (cn *conn) RemoteAddr() net.Addr {
 func (cn *conn) Close() error {
 	return cn.netcn.Close()
 }
+
+func isSameSlice(s1, s2 []byte) bool {
+	return len(s1) > 0 && len(s2) > 0 && &s1[0] == &s2[0]
+}
+
+func (cn *conn) copyBuf(b []byte) []byte {
+	if isSameSlice(b, cn.buf) {
```
Comment (vmihailenco) [inline conn.go]: This check is responsible for speedups.

File: parser.go [hunk 2]
```diff
@@ -227,74 +226,21 @@ func readLine(rd *bufio.Reader) ([]byte, error) {
...
-		if err != nil {
-			return nil, err
-		}
-
-		arg, err := readN(rd, int(argLen)+2)
-		if err != nil {
-			return nil, err
-		}
-		args = append(args, string(arg[:argLen]))
+func readN(cn *conn, n int) ([]byte, error) {
```
Comment (dim) [inline parser.go]: So I've been through this on another project (closed source) and we did pretty much exactly the same thing until it became a burden to manage. We then simply switched to the (less customisable) http://golang.org/pkg/bytes/#Buffer and benchmarks were on par.

**Comment by dim** [comment]
LGTM in general, but I also think that there is too much going on. A connection has a simple buffer which is filled and drained, seems like the perfect use case for `bytes.Buffer`

File: parser.go (hunk 2 above)
Comment (vmihailenco) [inline parser.go]: I am not sure I fully understand. How do you suggest to read data from `cn.rd *net.Conn` into `bytes.Buffer`? `buffer.ReadFrom` reads until `io.EOF` which does not work here.

Comment (dim) [inline parser.go]: Sorry, re-read. Ignore what I said, makes sense

**Pull request merged**